sd3_infer.py

```diff
@@ -111,34 +125,79 @@ def __init__(self):
 
 
 class T5XXL:
-    def __init__(self):
-        with safe_open("models/t5xxl.safetensors", framework="pt", device="cpu") as f:
-            self.model = T5XXLModel(T5_CONFIG, device="cpu", dtype=torch.float32)
-            load_into(f, self.model.transformer, "", "cpu", torch.float32)
+    def __init__(self, model_folder: str, device: str = "cpu", dtype=torch.float32):
+        with safe_open(
+            f"{model_folder}/t5xxl.safetensors", framework="pt", device="cpu"
+        ) as f:
+            self.model = T5XXLModel(T5_CONFIG, device=device, dtype=dtype)
+            load_into(f, self.model.transformer, "", device, dtype)
+
+
+CONTROLNET_MAP = {
+    "time_text_embed.timestep_embedder.linear_1.bias": "t_embedder.mlp.0.bias",
+    "time_text_embed.timestep_embedder.linear_1.weight": "t_embedder.mlp.0.weight",
+    "time_text_embed.timestep_embedder.linear_2.bias": "t_embedder.mlp.2.bias",
+    "time_text_embed.timestep_embedder.linear_2.weight": "t_embedder.mlp.2.weight",
+    "pos_embed.proj.bias": "x_embedder.proj.bias",
+    "pos_embed.proj.weight": "x_embedder.proj.weight",
+    "time_text_embed.text_embedder.linear_1.bias": "y_embedder.mlp.0.bias",
+    "time_text_embed.text_embedder.linear_1.weight": "y_embedder.mlp.0.weight",
+    "time_text_embed.text_embedder.linear_2.bias": "y_embedder.mlp.2.bias",
+    "time_text_embed.text_embedder.linear_2.weight": "y_embedder.mlp.2.weight",
+}
 
 
 class SD3:
-    def __init__(self, model, shift, verbose=False):
+    def __init__(
+        self, model, shift, control_model_file=None, verbose=False, device="cpu"
+    ):
+
+        # NOTE 8B ControlNets were trained with a slightly different forward pass and conditioning, 
+        # so this is a flag to enable that logic.
+        self.using_8b_controlnet = False
+
         with safe_open(model, framework="pt", device="cpu") as f:
+            control_model_ckpt = None
+            if control_model_file is not None:
+                control_model_ckpt = safe_open(
+                    control_model_file, framework="pt", device=device
+                )
             self.model = BaseModel(
                 shift=shift,
                 file=f,
                 prefix="model.diffusion_model.",
-                device="cpu",
+                device="cuda",
```

Comment:
The model is later loaded into cuda here : https://github.com/Stability-AI/sd3.5/blob/4e484e05308d83fb77ae6f680028e6c313f9da54/sd3_infer.py#L270, should this be cuda here?

Reply:
I think we want the dit to always be in cuda memory no? The text encoders we want to offload to `text_encoder_device`, but the diffusion model should stay in memory until program end